connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginUtilsTest.java

```diff
@@ -114,68 +90,264 @@ public void testConnectFrameworkClasses() {
         assertFalse(PluginUtils.shouldLoadInIsolation(
                 "org.apache.kafka.clients.admin.KafkaAdminClient")
         );
-        assertFalse(PluginUtils.shouldLoadInIsolation(
-                "org.apache.kafka.connect.rest.ConnectRestExtension")
-        );
     }
 
     @Test
-    public void testAllowedConnectFrameworkClasses() {
-        assertTrue(PluginUtils.shouldLoadInIsolation("org.apache.kafka.connect.transforms."));
-        assertTrue(PluginUtils.shouldLoadInIsolation(
-                "org.apache.kafka.connect.transforms.ExtractField")
-        );
-        assertTrue(PluginUtils.shouldLoadInIsolation(
-                "org.apache.kafka.connect.transforms.ExtractField$Key")
-        );
-        assertTrue(PluginUtils.shouldLoadInIsolation("org.apache.kafka.connect.json."));
-        assertTrue(PluginUtils.shouldLoadInIsolation(
-                "org.apache.kafka.connect.json.JsonConverter")
-        );
-        assertTrue(PluginUtils.shouldLoadInIsolation(
-                "org.apache.kafka.connect.json.JsonConverter$21")
-        );
-        assertTrue(PluginUtils.shouldLoadInIsolation("org.apache.kafka.connect.file."));
-        assertTrue(PluginUtils.shouldLoadInIsolation(
-                "org.apache.kafka.connect.file.FileStreamSourceTask")
-        );
-        assertTrue(PluginUtils.shouldLoadInIsolation(
-                "org.apache.kafka.connect.file.FileStreamSinkConnector")
-        );
+    public void testConnectApiClasses() {
+        String[] apiClasses = new String[] {
+            // Enumerate all packages and classes
+            "org.apache.kafka.connect.",
+            "org.apache.kafka.connect.components.",
+            "org.apache.kafka.connect.components.Versioned",
+            //"org.apache.kafka.connect.connector.policy.", isolated by default
+            "org.apache.kafka.connect.connector.policy.ConnectorClientConfigOverridePolicy",
+            "org.apache.kafka.connect.connector.policy.ConnectorClientConfigRequest",
+            "org.apache.kafka.connect.connector.",
+            "org.apache.kafka.connect.connector.Connector",
+            "org.apache.kafka.connect.connector.ConnectorContext",
+            "org.apache.kafka.connect.connector.ConnectRecord",
+            "org.apache.kafka.connect.connector.Task",
+            "org.apache.kafka.connect.data.",
+            "org.apache.kafka.connect.data.ConnectSchema",
+            "org.apache.kafka.connect.data.Date",
```

Comment:
> The exhaustive list of classes is unmaintainable

Yes, it will be difficult to keep this test perfectly up-to-date without due-diligence from reviewers and committers. I think this due-diligence is valuable, and will avoid a bug-fix PR like this being necessary in the future.

> it's highly improbable that new classes will be removed or added.

I would say infrequent, but not improbable. This next release includes two KIPs that added classes to the api, and each had to change the whitelist and/or tests (that's why this had merge conflicts earlier).

I don't think it's significantly more effort to maintain the exhaustive list than it is to maintain a reduced list of classes either. Consider this decision tree:

1. Did you add an API class? 
    If not -> It's not necessary to change this test, since it's only concerned with API classes.
2. Is your new class inside an existing isolated-by-default package?
    If so -> You'd need to exclude this class, and not doing so introduces a bug.
    Once you change PluginUtils, you'd naturally update PluginUtilsTest with your class name to verify the fix.
3. Is your new class in a new package
    If so -> You'd need to update this test even if it was a non-exhaustive packages-only test.

Only if you've fallen through all of those conditions, and added an API class in an existing package that is already whitelisted, would you save the time necessary to update this test. And in this case, the code is already correct, so the test silently diverges from the real class list, without any negative effects.

Reply:
Sorry, I wasn't clear. I meant that I think the bookkeeping of specific classes in this test might be hard and that we might miss adding or removing an entry once a class is added or removed in any of the packages. Totally agree that we add classes and interfaces to the API as needed. 

The basic principles behind classloading isolation was that we exclude everything in the framework from isolation and the users should not expect to package connect classes with their connectors. Yet, we had the need to introduce exceptions because we package some of the Connect plugins (Connectors, SMTs and Converters and etc) with the framework itself. That explains the inclusion and exclusion lists here and why specific classes/interfaces are present. The intention is keep this "per java package" approach and add only the necessary exceptions. That's why I suggested that an per class explicit listing here might be a bit tedious and perhaps a bit surprising to developers that contribute a class that is properly isolated but they'd be ask to add it here (e.g. a new SMT). 

Having said that the benefits of this PR outweigh my concern atm, so I'm happy to move forward and include the test refactoring as well.